File: 삼성코테/16236.py
# 0 : 빈칸
# 1-6: 칸에 있는 물고기의 크기
# 9: 아기 상어의 위치

# 아기 상어의 최초 크기 = 2
# 1초마다 상하좌우 중 한칸 이동
# 자신보다 큰 물고기가 있는 칸을 지나갈 수 없음
# 자기보다 작은 크기의 물고기만 먹을 수 있음
# 크기 같은 물고기가 있는 곳은 지나갈 수는 있으나 먹을 수 없음

from collections import deque

N = int(input())
maze = []
start = []
time = 0

fish = []
size = 2
eat = 0

for i in range(N):
    row = list(map(int, input().split()))

    for j in range(len(row)):
        if row[j] == 9:
            start.append([i, j])
        elif 0 < row[j] <= 6:
            fish.append(row[j])
    maze.append(row)

if len(start) == 0:
    print(0)

# 먹을 수 있는 물고기가 미존재
if len(fish) == 0:
    print(0)

# 물고기가 최소 1개이상 존재하는 경우
else:
    fish.sort()

    # 물고기는 있지만 먹을 수 없는 것들만 있을 때
    if fish[0] >= size:
        print(0)

    # 먹을 수 있는 물고기가 있을 경우
    else:
        def isOK(a, b, width, height):
            return 0 <= a < height and 0 <= b < width


        # 먹을 물고기가 없다면 어미에게 헬프콜
        # 1마리 먹을 수 있으면 그 물고기 먹으러 이동
        # 1마리보다 먹을 수 있는 물고기 수가 많다면 가장 가까운 물고기를 먹으러 간다
        # 위쪽 -> 왼쪽 순서로 우선순위 발동
        # 자기 자신의 크기와 같은 수의 물고기 먹을 때마다 크기가 1 증가

        # 상-좌-우-하
        dir_x = [-1, 0, 0, 1]
        dir_y = [0, -1, 1, 0]
        index = [[0] * N for _ in range(N)]

        # [[[]]]
        queue = deque([start])
        index[start[0][0]][start[0][1]] = 1
        maze[start[0][0]][start[0][1]] = 0
        count = 0

        while queue:
            # [[]]
            target = queue.popleft()
            # []
            box = []
            # 먹이섭취여부
            found = 0
            food = []

            for i in target:
                # 위로이동
                for p in range(4):
                    # 타당한 좌표, 미방문여부 확인
                    if isOK(i[0]+dir_x[p], i[1]+dir_y[p], N, N) and index[i[0] + dir_x[p]][i[1] + dir_y[p]] == 0:
                        # 좌표 값이 0 또는 자신의 크기와 같을 경우 (통행만 가능한 경우)
                        if maze[i[0] + dir_x[p]][i[1] + dir_y[p]] == 0 or maze[i[0] + dir_x[p]][i[1] + dir_y[p]] == size:
                            # 일단 좌표값을 임시저장 + 방문여부 표시
                            box.append([i[0] + dir_x[p], i[1] + dir_y[p]])
                            index[i[0] + dir_x[p]][i[1] + dir_y[p]] = 1
                        # 자신보다 크기가 작은 물고기가 있는 칸인 경우
                        # 해당좌표 값만 queue에 남기고, 식사 횟수 증가 및 크기 확인
                        elif 0 < maze[i[0] + dir_x[p]][i[1] + dir_y[p]] < size:
                            food.append([i[0] + dir_x[p], i[1] + dir_y[p]])
                            index[i[0] + dir_x[p]][i[1] + dir_y[p]] = 1
                            # 먹이찾았음을 표시
                            found = 1

            # 먹이를 발견했을 경우
            if found == 1:
                # 먹이 발견한 곳이 여러곳이라면
                if len(food) > 1:
                    # 높이기준으로 오름차순 정렬
                    food.sort(key=lambda x: x[0])

                eat += 1
                fish.remove(maze[food[0][0]][food[0][1]])
                maze[food[0][0]][food[0][1]] = 0
                food = food[:1]
                count += 1
                time += count
                count = 0

                # 자신의 사이즈만큼 물고기를 먹은 상태라면 크기 증가
                if eat == size:
                    size += 1
                    eat = 0

                # 물고기 다먹었다면
                if len(fish) == 0:
                    break

                # 먹을 수 있는 물고기가 없다면
                elif min(fish) >= size:
                    break

                # 먹을 수 있는 물고기가 존재한다면
                else:
                    queue = deque([food])
                    index = [[0] * N for _ in range(N)]
                    index[food[0][0]][food[0][1]] = 1

            # 먹이를 발견하지 못한 경우
            elif found == 0:
                # 이동이 가능한 경우
                if len(box) > 0:
                    box.sort(key=lambda x: x[0])
                    queue.append(box)
                    count += 1

        print(time)

File: 삼성코테/16236.java
import java.util.*;
import java.io.*;
public class Baek_16236 {
    static int[][] maze;
    static int start_x;
    static int start_y;
    static int count = 0;
    static int level = 2;
    static int answer = 0;
    static int N;
    static boolean [][] visited;
    static class Spot implements Comparable<Spot> {
        int x;
        int y;
        int level;
        int distance;
        Spot(int x, int y, int level) {
            this.x = x;
            this.y = y;
            this.level = level;
            this.distance = 0;
        }
        public void updateDistance(int distance) {
            this.distance = distance;
        }

        @Override
        public int compareTo(Spot o) {
            if(this.distance < o.distance)
                return -1;
            else if(this.distance > o.distance)
                return 1;
            else {
                if(this.x < o.x)
                    return -1;
                else if(this.x > o.x)
                    return 1;
                else {
                    if(this.y > o.y)
                        return 1;
                    else
                        return -1;
                }
            }
        }
    }
    static ArrayList<int[]> lis = new ArrayList<>();
    public static boolean findFish() {
        PriorityQueue<Spot> queue = new PriorityQueue<>();

        for(int[] fish: lis) {
            if(level > fish[2]) {
                int distance = getCost(fish[0], fish[1]);
                if(distance != -1) {
                    Spot eat = new Spot(fish[0], fish[1], fish[2]);
                    eat.updateDistance(distance);
                    queue.add(eat);
                }
            }
        }

        int removeIdx = 0;
        if(!queue.isEmpty()) {
            Spot fish = queue.peek();
            answer += fish.distance;
            maze[fish.x][fish.y] = 0;
            start_x = fish.x;
            start_y = fish.y;
            ++count;

            if(count == level) {
                count = 0;
                ++level;
            }

            for (int i = 0; i < lis.size(); i++) {
                if (lis.get(i)[0] == queue.peek().x && lis.get(i)[1] == queue.peek().y) {
                    removeIdx = i;
                    break;
                }
            }
            lis.remove(removeIdx);
            return true;
        }

        return false;
    }
    public static boolean isOk(int a, int b) {
        return (0 <= a && a < N && 0 <= b && b < N);
    }
    public static int getCost(int a, int b) {
        int [] dir_x = {-1, 0, 1, 0};
        int [] dir_y = {0, 1, 0, -1};
        boolean [][] temp = new boolean[N][N];
        temp[start_x][start_y] = true;

        Queue<Spot> info = new LinkedList<>();
        info.add(new Spot(start_x, start_y, level));

        while(!info.isEmpty()) {
            Spot now = info.poll();
            for (int i = 0; i < 4; i++) {
                int dx = now.x + dir_x[i];
                int dy = now.y + dir_y[i];

                if(dx == a && dy == b) {
                    visited[dx][dy] = true;
                    return now.distance + 1;
                }

                if (isOk(dx, dy)) {
                    if ((maze[dx][dy] <= level) && !temp[dx][dy]) {
                        Spot adder = new Spot(dx, dy, level);
                        adder.updateDistance(now.distance + 1);
                        info.add(adder);
                        temp[dx][dy] = true;
                    }
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        maze = new int[N][N];
        visited = new boolean[N][N];

        StringTokenizer st;
        for(int i=0; i<N; i++) {
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<N; j++) {
                int num = Integer.parseInt(st.nextToken());
                if(num == 9) {
                    start_x = i;
                    start_y = j;
                } else if(num != 0) {
                    lis.add(new int[]{i, j, num});
                    maze[i][j] = num;
                }
            }
        }

        visited[start_x][start_y] = true;

        if(lis.isEmpty())
            System.out.println(0);
        else if(lis.size() == 1) {
            if(lis.get(0)[2] < 2)
                System.out.println(Math.abs(lis.get(0)[0] - start_x) + Math.abs(lis.get(0)[1] - start_y));
            else
                System.out.println(0);
        } else {
            while (true) {
                boolean result = findFish();
                if (!result)
                    break;
            }
            System.out.println(answer);
        }
    }
}
